File: 2022/day02/solutionB.rb
#!/bin/env ruby

SCORES = ['A', 'B', 'C']
WINS = ['A B', 'C A', 'B C']
DRAWS = ['A A', 'B B', 'C C']
TO_LOSE = ['C', 'A', 'B']
TO_WIN = ['B', 'C', 'A']

input = File.read('./input.txt')

# @param {String} input
# @return {Array(String)}
def format_input(input)
  input.split("\n")
end

# @param {Array(String)} formatted_input
# @return {Integer}
def solution(formatted_input)
  formatted_input.map do |round|
    if round[2] == 'X'
      round[2] = TO_LOSE[SCORES.index round[0]]
    elsif round[2] == 'Y'
      round[2] = round[0]
    elsif round[2] == 'Z'
      round[2] = TO_WIN[SCORES.index round[0]]
    end
    
    score = 1 + SCORES.index(round[2])

    if WINS.include? round
      score += 6
    elsif DRAWS.include? round
      score += 3
    end
    
    score
  end.sum
end

print solution(format_input(input))

File: 2022/day02/python/solutionB.py
#!/bin/env python

SCORES = ['A', 'B', 'C']
CASES = ['X', 'Y', 'Z']
WINS = ['A B', 'C A', 'B C']
DRAWS = ['A A', 'B B', 'C C']
TO_LOSE = ['C', 'A', 'B']
TO_WIN = ['B', 'C', 'A']

input = open('../input.txt').read()

def format(input: str) -> list[str]:
    return input.split('\n')

def solution(rounds: list[str]) -> int:
    def _map(round):
        [opponent, _, player] = round
        cases = [TO_LOSE[SCORES.index(opponent)], opponent, TO_WIN[SCORES.index(opponent)]]
        player = cases[CASES.index(player)]
        round = f"{opponent} {player}"

        score = 1 + SCORES.index(player)

        if round in WINS:
            score += 6
        elif round in DRAWS:
            score += 3

        return score
        
    return sum(map(_map, rounds))

print(solution(format(input)))
